Смышляев Дмитрий

vafle228 opened the pull request.

vafle228 committed:
Добавил PersonBuilder
- Добавил билдера
- Встроил ему методы по созданию царя
- Встроил метод по созданию родителей царя

vafle228 committed:
Пояснил за позор
- Написал комментарий к альтернативному решению

vafle228 committed:
Рефактор валидации
- Разнес тесты для конструктора
- Параметризовал тесты для IsValidNumber метода
- Добавил парочку новых тестов

vafle228 committed:
Merge pull request #1 from vafle228/ObjectComparisonRefactor

Object comparison refactor

vafle228 committed:
Merge pull request #2 from vafle228/NumberValidatorRefactor

Рефактор валидации

w1jtoo commented on Testing/Basic/Homework/1. ObjectComparison/ObjectComparison.cs:
```diff
@@ -9,27 +10,36 @@ public class ObjectComparison
+         * 2. Метод AreEqual проверяет ограниченное множество полей. Представим, что мы хотим добавить
+         * новое поле для объекта Person. В таком случаи придется дописывать условия, для проверки нового поля
+         *
+         * 3. Пожалуй самое страшное. Если класс Person будет содержать любой другой объект, кроме Person, то
```
Тут скорее не соглашусь. Потому что тот подход, который выбрал ты тоже не лишен такого минуса. Ситуация: если бы Person содержал список актуальных контактов, представленный сложной структурой данных. Как их сравнивать? Вообще нетривиальный аспект.

На это сообщение можешь не отвечать, просто резолви этот тред, если нечего добавить.

vafle228 replied:
Проблема с вложенными объектами будет только если
1) У нас глубокое вложение (больше 10, если судить по доке)
2) У нас будет пул полей для проверки (тут таким пулом стали все поля без айди)

В таком случаи действительно возникнет необходимость дописать код, чтобы тесты работали корректно. Но моя позиция все еще справедлива, так как для легких изменений (типа добавить город проживания) тест не придется переписывать. А в случаях сложного изменения, усилия затраченные на исправления моего теста будут меньше, чем на исправления метода AreEqual (более подробно в пункте 3)

vafle228 replied:
Типо понятно что можно придумать сколько угодно сложное поле, которое вынудит внести правки в мой тест, но повторюсь, учесть то же самое поле методом AreEqual будет кратно сложнее

w1jtoo commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,57 @@
+    private const string SCALE_ERROR_MESSAGE = "precision must be a non-negative number less or equal than precision";
+    
+    # region NumberValidator constructor tests
+    [TestCase(-1, 2, PRECISION_ERROR_MESSAGE, TestName = "precision <= 0")]
```
Прям в `TestCase` есть параметр `Description`. По логике он лучше сюда подходит

w1jtoo commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -1,31 +1,57 @@
+    }
+    # endregion NumberValidator constructor tests
+    
+    # region NumberValidator IsValidNumber method tests
```
Есть небольшие нарекания по понимаю этих тестов. 

Они правильные, спросу нет. Но вот их оправдание и ответ на вопрос "почему так" получить не получится. Только после прочтения `NumberValidator`, а мы хотели бы этого избежать.

Моё предложение разделить описанные тобой кейсы логически на группы. Их группировать по названиям теста и, использовать TestCase для нескольких наборов данных. Как раз название теста использовать для ответа на вопрос "почему так"

vafle228 committed:
Разбил тесты на логические подгруппы

vafle228 committed:
Переосмысление одного теста
- Убрал тест на проверку знака при нуле
- Перенес все тест кейсы с убранного теста в тест на укладывание в точность
